### PR TITLE
removed the user's system language as the default content language

### PR DESCRIPTION
**Pull request by marcelmoosbrugger**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | yes
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | https://github.com/massiveart/husky/pull/695
| License | MIT
| Documentation PR | none

#### What's in this PR?

In almost every section of Sulu the user's system locale was used as the fallback if no locale was given. This PR removes this behavior in sections where this is not appropriate.

#### Why?

Recently the system locales got removed as content languages. That means if a locale is just for the backend and there is no webspace etc. which also uses this locale, it is not (and should not be) possible to create content in this locale. The fact that everywhere the user's system locale was used as a fallback, made Sulu inoperative when such a locale was used in the backend.

### CODE REVIEW

**Comment by marcelmoosbrugger** [comment]
@danrot. This PR is ready for review :) Moreover the behat test work for me locally.

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/header/main.js [hunk 1]
```diff
@@ -501,13 +501,15 @@ define([], function() {
         startLanguageChanger: function() {
             if (!!this.options.toolbarLanguageChanger) {
                 var $element = this.sandbox.dom.createElement(this.sandbox.util.template(templates.languageChanger)({
-                        title: this.options.toolbarLanguageChanger.preSelected || this.sandbox.sulu.user.locale
+                            title: this.options.toolbarLanguageChanger.preSelected
+                                || this.sandbox.sulu.getDefaultContentLocale()
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/header/main.js]: Isn't that the exact error we had before? This method is simply returning the first localization in the system again, isn't it?

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/aura_extensions/sulu-extension.js [hunk 2]
```diff
@@ -210,6 +210,19 @@
...
+             * user and the available locales.
+             *
+             * @returns {String} the default locale
+             */
+            app.sandbox.sulu.getDefaultContentLocale = function() {
+                if (!!SULU.user.locale && _.contains(SULU.locales, SULU.user.locale)) {
+                    return SULU.user.locale;
+                }
+
+                return SULU.locales[0];
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/aura_extensions/sulu-extension.js]: Are these locales only the content locales?

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js [hunk 3]
```diff
@@ -163,7 +163,7 @@ define([
                     '/admin/api/media/' + curMediaId + '?action=new-version' :
                     '/admin/api/media?collection=' + this.formOptions.accountAvatarCollection;
                 
-                url = url + '&locale=' + encodeURIComponent(this.sandbox.sulu.user.locale);
+                url = url + '&locale=' + encodeURIComponent(this.sandbox.sulu.getDefaultContentLocale());
```
Comment (danrot) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js]: Just returning the first content locale is probably not a good solution here... Maybe we should also take the content language into account? /cc @chirimoya

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/contacts/edit/details/main.js [hunk 4]
```diff
@@ -152,7 +152,7 @@ define([
                     '/admin/api/media/' + curMediaId + '?action=new-version' :
                     '/admin/api/media?collection=' + this.formOptions.contactAvatarCollection;
 
-                url = url + '&locale=' + encodeURIComponent(this.sandbox.sulu.user.locale);
+                url = url + '&locale=' + encodeURIComponent(this.sandbox.sulu.getDefaultContentLocale());
```
Comment (danrot) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/contacts/edit/details/main.js]: Same as above

File: src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php [hunk 5]
```diff
@@ -111,4 +112,12 @@ private function getDocumentManager()
     {
         return $this->get('sulu_document_manager.document_manager');
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLocale(Request $request)
+    {
+        return;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php]: This controller uses locales, doesn't it?

File: src/Sulu/Bundle/ContentBundle/Controller/SeoController.php [hunk 6]
```diff
@@ -117,4 +117,12 @@ public function postAction(Request $request, $uuid)
             $this->view($result)
         );
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLocale(Request $request)
+    {
+        return $this->getLanguage($request);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Controller/SeoController.php]: Since the `getLanguage` method is private anyway I would remove it, and move the code here.

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js (hunk 3 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js]: It does not always return the first content locale. If the system locale of the current user is also a content locale, the current system locale is returned. But your right, a better solution would be to save the default content locale for a user.

File: src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js [hunk 7]
```diff
@@ -83,12 +83,13 @@ define([
...
             var security = Config.get('sulu_security.contexts')['sulu.webspace_settings.' + this.data.key + '.custom-urls'],
                 buttons = {},
+                locale = this.sandbox.sulu.getDefaultContentLocale(),
                 components = [
                     {
                         name: 'datagrid@husky',
                         options: {
                             el: '#webspace-custom-url-list',
-                            url: CustomUrlManager.generateUrl(this.data, null, null),
+                            url: CustomUrlManager.generateUrl(this.data, null, null) + '?locale=' + locale,
```
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js]: Why do we need the `locale` here? Don't we use the `targetLocale` from the body here?

File: src/Sulu/Bundle/CustomUrlBundle/Tests/Functional/Controller/CustomUrlControllerTest.php [hunk 8]
```diff
@@ -753,7 +753,7 @@ public function testCDelete($items)
         $this->assertHttpStatusCode(204, $response);
 
         $client = $this->createAuthenticatedClient();
-        $client->request('GET', '/api/webspaces/sulu_io/custom-urls');
+        $client->request('GET', '/api/webspaces/sulu_io/custom-urls?locale=en');
```
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Tests/Functional/Controller/CustomUrlControllerTest.php]: Could be removed if the above comment is true.

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php [hunk 9]
```diff
@@ -368,14 +369,16 @@ protected function saveEntity($id, Request $request)
             $collectionManager = $this->getCollectionManager();
             $data = $this->getData($request);
             $data['id'] = $id;
-            $data['locale'] = $this->getLocale($request);
+            $data['locale'] = $this->getRequestParameter($request, 'locale', true);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: Why don't you use the `getLocale` method anymore resp. override it to match the collections locales?

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php [hunk 10]
```diff
@@ -368,14 +369,16 @@ protected function saveEntity($id, Request $request)
...
             $collection = $collectionManager->save($data, $this->getUser()->getId());
 
             $view = $this->view($collection, 200);
         } catch (CollectionNotFoundException $cnf) {
             $view = $this->view($cnf->toArray(), 404);
         } catch (MediaException $me) {
             $view = $this->view($me->toArray(), 400);
+        } catch (MissingParameterException $mp) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: We call the exception variables in the `catch` block `$e`, please also fix for the other catch statements.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/edit/files/main.js [hunk 11]
```diff
@@ -231,10 +231,11 @@ define([
          * Start the list toolbar and the datagrid
          */
         startDatagrid: function() {
-            var view = UserSettingsManager.getMediaListView();
+            var view = UserSettingsManager.getMediaListView(),
+                locale = UserSettingsManager.getMediaLocale(),
 
             // init list-toolbar and datagrid
-            var settingsDropdown = [], buttons = {};
+            settingsDropdown = [], buttons = {};
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/edit/files/main.js]: wrong indentation

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js [hunk 12]
```diff
@@ -87,6 +87,7 @@ define([
          */
         initialize: function() {
             // extend defaults with options
+            defaults.locale = this.sandbox.sulu.getDefaultContentLocale();
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js]: Wouldn't it be better to pass the `locale` in the object of the second parameter of the `extend`? Then the `defaults` object wouldn't have to take state.

File: src/Sulu/Bundle/ResourceBundle/Resources/public/js/components/filters/components/form/main.js [hunk 13]
```diff
@@ -136,7 +136,7 @@ define(['config'], function(Config) {
                     options: {
                         el: $element,
                         fieldsUrl: typeConfig.fields,
-                        operatorsUrl: constants.operatorsUrl,
+                        operatorsUrl: constants.operatorsUrl + '?locale=' + AppConfig.getUser().locale,
```
Comment (danrot) [inline src/Sulu/Bundle/ResourceBundle/Resources/public/js/components/filters/components/form/main.js]: Do we still have to use the user locale here? That's because the filters themselves are not localized, right?

Comment (danrot) [inline src/Sulu/Bundle/ResourceBundle/Resources/public/js/components/filters/components/form/main.js]: No, the name of the filter is translated, right? In which language? I think you are also not allowed to use the locale here.

File: src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js (hunk 7 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js]: The cgetAction expects a locale to load the custom urls for. The part of the JS you commented is responsible for loading the list. There is no target locale available.

File: src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php (hunk 9 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Controller/CollectionController.php]: That's because at this point the locale is really required and the process should fail if its not given. In other methods the locale is not necessarily required and can also be null, which is the default behavior of `getLocale` in the RestController.

File: src/Sulu/Bundle/ResourceBundle/Resources/public/js/components/filters/components/form/main.js (hunk 13 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ResourceBundle/Resources/public/js/components/filters/components/form/main.js]: The operators are translated in all three system languages (and even some more if I can remember). In my opinion the operators need to be translated in the system language of the user, because it's weird to see everything in english but the translations of the operators are in lets say french.

File: CHANGELOG.md [hunk 14]
```diff
@@ -2,6 +2,7 @@ CHANGELOG for Sulu
 ==================
 
 * dev-develop
+<<<<<<< 0418309445ca627df6408cf9b216169ca035e6de
```
Comment (danrot) [inline CHANGELOG.md]: Remove this merge conflict

File: UPGRADE.md [hunk 15]
```diff
@@ -2,6 +2,18 @@
 
 ## 1.3.0-RC2
 
+### RestController locale
+
+The `getLocale` method of the RestController returned the locale
+of the user (a system locale), if no request parameter with the name
+`locale` was passed. As RestControllers most often provide access
+to content in content locales and not in system locales,
+this behaviour got removed. The `getLocale` method now just returns
```
Comment (danrot) [inline UPGRADE.md]: `was removed`

File: src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js [hunk 16]
```diff
@@ -93,7 +93,8 @@ define(function () {
             this.sandbox.dom.html(this.$el, this.renderTemplate('/admin/category/template/category/list'));
 
             // init list-toolbar and datagrid
-            this.sandbox.sulu.initListToolbarAndList.call(this, 'categories', '/admin/api/categories/fields',
+            this.sandbox.sulu.initListToolbarAndList.call(this,
+                'categories', '/admin/api/categories/fields',
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: own line for each parameter

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php [hunk 17]
```diff
@@ -173,7 +173,7 @@ public function testGetById()
 
         $client->request(
             'GET',
-            '/api/categories/' . $this->category1->getId()
+            '/api/categories/' . $this->category1->getId() . '?locale=en'
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: `locale` is a mandatory parameter now, right? Can we also add a test which checks if the request fails, when the locale is not passed?

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js (hunk 3 above)
Comment (danrot) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js]: But if the user locale is not also a content locale we are running into troubles here... When the first content language is one the user doesn't understand there is also no way for him to change that. Using the content locale before using the default locale is probably the better choice here.

File: src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php [hunk 18]
```diff
@@ -31,14 +32,14 @@ class NodeResourcelocatorController extends RestController implements ClassResou
      *
      * @return \Symfony\Component\HttpFoundation\Response
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php]: Please change to short name

File: src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php [hunk 19]
```diff
@@ -95,7 +95,7 @@ public function postAction($webspaceKey, Request $request)
         $document = $this->get('sulu_custom_urls.manager')->create(
             $webspaceKey,
             $request->request->all(),
-            $this->getLocale($request)
+            $this->getRequestParameter($request, 'targetLocale', true)
```
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php]: Does the `getLocale` method matter in this file? Then it should be overridden with returning null.

File: src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js (hunk 7 above)
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js]: Ok, now I see what this is for... Actually it should also use the content language of the current user here...

File: src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php (hunk 19 above)
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php]: The `getLocale` method should actually return null, and at the same time the `locale` must be a mandatory parameter for the `cgetAction` I'd say. Then everybody can access the page without weird locale permission errors, and at the same time the title for the list would be correct.

Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php]: This would also mean that there has to be a new test for the `cgetAction`.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js [hunk 20]
```diff
@@ -87,7 +86,9 @@ define([
          */
         initialize: function() {
             // extend defaults with options
-            this.options = this.sandbox.util.extend(true, {}, defaults, this.options);
+            this.options = this.sandbox.util.extend(true, {
+                locale: this.sandbox.sulu.getDefaultContentLocale
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js]: Arent you missing `()` at the end here?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/collection-select-overlay/main.js [hunk 21]
```diff
@@ -75,7 +74,9 @@ define(function() {
          */
         initialize: function() {
             // extend defaults with options
-            this.options = this.sandbox.util.extend(true, {}, defaults, this.options);
+            this.options = this.sandbox.util.extend(true, {
+                locale: this.sandbox.sulu.getDefaultContentLocale
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/collection-select-overlay/main.js]: Aren't you missing `()` at the end here?

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php [hunk 22]
```diff
@@ -888,6 +889,7 @@ public function testPostWithSmallFile()
             'POST',
             '/api/media',
             [
+                'locale' => 'en',
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php]: Is that also mandatory now in the `MediaController`? Then a test case for that is probably missing.

File: src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php [hunk 23]
```diff
@@ -588,7 +593,7 @@ public function testCDeleteByIdsPartialExistent()
      */
     public function testDeleteByIdNotExisting()
     {
-        $this->client->request('GET', '/api/filters/666');
+        $this->client->request('GET', '/api/filters/666?locale=de');
```
Comment (danrot) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php]: Also add test case for failing request without `locale`.

File: src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/OperatorControllerTest.php [hunk 24]
```diff
@@ -133,7 +133,7 @@ public function testCget()
     {
         $this->client->request(
             'GET',
-            '/api/operators'
+            '/api/operators?locale=de'
```
Comment (danrot) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/OperatorControllerTest.php]: Testcase for missing locale

File: src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/UserControllerTest.php [hunk 25]
```diff
@@ -523,7 +523,7 @@ public function testPut()
 
         $client->request(
             'GET',
-            '/api/users/' . $this->user1->getId()
+            '/api/users/' . $this->user1->getId() . '?locale=en'
```
Comment (danrot) [inline src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/UserControllerTest.php]: Why is that necessary? You didn't change this controller.

File: UPGRADE.md (hunk 15 above)
Comment (marcelmoosbrugger) [inline UPGRADE.md]: fixed

File: src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js (hunk 16 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CategoryBundle/Resources/public/js/components/categories/list/main.js]: fixed

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php (hunk 17 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: In the CategoryController the locale is not mandatory. When no locale is passed, the categories are returned in their fallback locale.

File: src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php (hunk 18 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ContentBundle/Controller/NodeResourcelocatorController.php]: fixed

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js (hunk 20 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/media-edit-overlay/main.js]: fixed.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/collection-select-overlay/main.js (hunk 21 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/collections/collection-select-overlay/main.js]: fixed.

File: src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/UserControllerTest.php (hunk 25 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/UserControllerTest.php]: fixed

File: src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php (hunk 19 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php]: I'd say that the locale is mandatory, as both getAction and cgetAction fail with 500 when no locale is passed. I'll add a test for both of them.

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js (hunk 3 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/accounts/edit/details/main.js]: fixed

File: src/Sulu/Bundle/ContactBundle/Resources/public/js/components/contacts/edit/details/main.js (hunk 4 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ContactBundle/Resources/public/js/components/contacts/edit/details/main.js]: fixed

File: src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js (hunk 7 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CustomUrlBundle/Resources/public/js/components/webspace/settings/custom-url/main.js]: fixed

File: src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php (hunk 19 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CustomUrlBundle/Controller/CustomUrlController.php]: fixed.

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php (hunk 22 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php]: Media are returned in the fallback locale when no locale is specified.

File: src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php (hunk 23 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php]: The request for a filter does not fail when no locale is provided.

File: src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/OperatorControllerTest.php (hunk 24 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/OperatorControllerTest.php]: Like with the FilterController, the request does not fail, when no locale is provided.

Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/OperatorControllerTest.php]: I made the locale mandatory and added some tests.

File: src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php (hunk 23 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ResourceBundle/Tests/Functional/Controller/FilterControllerTest.php]: I made the locale mandatory and added some tests.

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php (hunk 22 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php]: I made the locale mandatory and added some tests.

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php (hunk 17 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: I made the locale mandatory and added some tests.

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 26]
```diff
@@ -370,6 +384,14 @@ public function getSecurityContext()
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function getLocale(Request $request)
+    {
+        return;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Shouldn't this be returning something? Otherwise the security won't apply here, but I think it certainly should (because media is translatable)

**Comment by danrot** [comment]
When you open a contact, the request for the categories fails, and you get an infinite loader.

**Comment by danrot** [comment]
CustomUrls cannot be edited, because when clicking on them in the datagrid a 400 response is returned.

**Comment by danrot** [comment]
DefaultSnippets cannot be edited.

**Comment by danrot** [comment]
The categories do also not work for the content type and excerpt tab.

**Comment by danrot** [comment]
The language of a user cannot be changed, because the `UserController` also uses the implementation of the `RestController`, and `$request->get()` also takes the request body into account.

**Comment by marcelmoosbrugger** [comment]
@danrot the issues should be fixed now
